bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/BalRuntime.java

```diff
@@ -66,55 +69,66 @@ public class BalRuntime extends Runtime {
+        schedulerThread = new Thread(scheduler::start);
+        schedulerThread.start();
+        invokeMethodSync("$moduleInit");
         moduleInitialized = true;
     }
 
     public void start() {
         if (!moduleInitialized) {
-            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_METHOD_CALL, "start");
+            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_FUNCTION_INVOCATION_BEFORE_MODULE_INIT, "start");
+        }
+        if (moduleStarted) {
+            throw ErrorHelper.getRuntimeException(ErrorCodes.FUNCTION_ALREADY_CALLED, "start");
         }
-        invokeMethodAsync("$moduleStart", null, PredefinedTypes.TYPE_NULL, "start", new Object[1]);
+        invokeMethodSync("$moduleStart");
+        moduleStarted = true;
     }
 
     public void invokeMethodAsync(String functionName, Callback callback, Object... args) {
         if (!moduleInitialized) {
-            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_FUNCTION_INVOCATION, functionName);
+            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_FUNCTION_INVOCATION_BEFORE_MODULE_INIT,
+                    functionName);
         }
-        invokeMethodAsync(functionName, callback, PredefinedTypes.TYPE_ANY, functionName, args);
+        invokeMethod(functionName, callback, PredefinedTypes.TYPE_ANY, functionName, args);
     }
 
     public void stop() {
         if (!moduleInitialized) {
-            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_METHOD_CALL, "stop");
+            throw ErrorHelper.getRuntimeException(ErrorCodes.INVALID_FUNCTION_INVOCATION_BEFORE_MODULE_INIT, "stop");
         }
-        invokeMethodAsync("$moduleStop", null, PredefinedTypes.TYPE_NULL, "stop", new Object[1]);
-    }
-
-    private void invokeMethodAsync(String functionName, Callback callback, Type returnType, String strandName,
-                                   Object... args) {
-        ValueCreator valueCreator = ValueCreator.getValueCreator(ValueCreator.getLookupKey(module.getOrg(),
-                module.getName(), module.getMajorVersion(), module.isTestPkg()));
-        Function<?, ?> func = o -> valueCreator.call((Strand) (((Object[]) o)[0]), functionName, args);
-        FutureValue future = scheduler.createFuture(null, callback, null, returnType, strandName, null);
-        Object[] argsWithStrand = new Object[args.length + 1];
-        argsWithStrand[0] = future.strand;
-        System.arraycopy(args, 0, argsWithStrand, 1, args.length);
-        scheduler.schedule(argsWithStrand, func, future);
-        scheduler.start();
+        if (moduleStopped) {
+            throw ErrorHelper.getRuntimeException(ErrorCodes.FUNCTION_ALREADY_CALLED, "stop");
+        }
+        scheduler.poison();
+        try {
+            schedulerThread.join();
+        } catch (InterruptedException e) {
+            throw ErrorCreator.createError(StringUtils.fromString("ballerina: error occurred in while waiting for " +
```

Comment:
```suggestion
            throw ErrorCreator.createError(StringUtils.fromString("ballerina: error occurred while waiting for " +
```
This might be better